File: src/services/report.js
/**
 * Formatting helpers that turn validated result objects into the strings
 * offered for download. Pure functions. Direct port of ui/report.py.
 */

/** Build export Markdown without escaping task text, preserving inline links. */
function roadmapToMarkdown(roadmap) {
  if (!roadmap || !roadmap.weeks || !roadmap.weeks.length) {
    return "_No AI roadmap was generated for this assessment._";
  }

  const lines = [];
  for (const week of roadmap.weeks) {
    lines.push(`### ${week.week_title}`);
    for (const task of week.tasks) {
      lines.push(`- [ ] ${task}`);
    }
    lines.push("");
  }

  return lines.join("\n").trim();
}

function buildMarkdownReport(role, results, roadmapText) {
  const lines = [
    `# SkillGap Assessment Report: ${role}`,
    "",
    `**Experience Level:** ${results.experience_level}`,
    "",
    `**Role Readiness:** ${results.readiness_score.toFixed(0)}%`,
    `**Skill Coverage:** ${results.coverage_score.toFixed(0)}%`,
    "",
    "## Strong Matches",
  ];
  if (results.strong_matches && results.strong_matches.length) {
    lines.push(...results.strong_matches.map((item) => `- ${item}`));
  } else {
    lines.push("- None yet.");
  }

  lines.push("");
  lines.push("## Needs Improvement");
  if (results.needs_improvement && results.needs_improvement.length) {
    lines.push(...results.needs_improvement.map((item) => `- ${item}`));
  } else {
    lines.push("- Nothing stuck at Beginner level.");
  }

  lines.push("");
  lines.push("## Critical Gaps");
  if (results.critical_missing && results.critical_missing.length) {
    lines.push(...results.critical_missing.map((req) => `- ${req}`));
  } else {
    lines.push("- None. Full coverage.");
  }

  lines.push("");
  lines.push("## 4-Week AI Roadmap");
  lines.push("");
  lines.push(roadmapText || "_No AI roadmap was generated for this assessment._");

  return lines.join("\n");
}

module.exports = { roadmapToMarkdown, buildMarkdownReport };

File: ui/report.py
"""Formatting helpers that turn validated schema objects into the strings the
UI displays or offers for download. Pure functions, no Streamlit calls."""
from __future__ import annotations

from typing import Optional

from services.schema import AnalysisResult, RoadmapResponse


def roadmap_to_markdown(roadmap: Optional[RoadmapResponse]) -> str:
    """Build export Markdown without escaping task text, preserving inline links."""
    if not roadmap or not roadmap.weeks:
        return "_No AI roadmap was generated for this assessment._"

    lines = []
    for week in roadmap.weeks:
        lines.append(f"### {week.week_title}")
        for task in week.tasks:
            lines.append(f"- [ ] {task}")
        lines.append("")

    return "\n".join(lines).strip()


def build_markdown_report(role: str, results: AnalysisResult, roadmap_text: str) -> str:
    lines = [
        f"# SkillGap Assessment Report: {role}",
        "",
        f"**Experience Level:** {results.experience_level}",
        "",
        f"**Role Readiness:** {results.readiness_score:.0f}%",
        f"**Skill Coverage:** {results.coverage_score:.0f}%",
        "",
        "## Strong Matches",
    ]
    if results.strong_matches:
        lines.extend(f"- {item}" for item in results.strong_matches)
    else:
        lines.append("- None yet.")

    lines.append("")
    lines.append("## Needs Improvement")
    if results.needs_improvement:
        lines.extend(f"- {item}" for item in results.needs_improvement)
    else:
        lines.append("- Nothing stuck at Beginner level.")

    lines.append("")
    lines.append("## Critical Gaps")
    if results.critical_missing:
        lines.extend(f"- {req}" for req in results.critical_missing)
    else:
        lines.append("- None. Full coverage.")

    lines.append("")
    lines.append("## 4-Week AI Roadmap")
    lines.append("")
    lines.append(roadmap_text if roadmap_text else "_No AI roadmap was generated for this assessment._")

    return "\n".join(lines)
